modules/nat64/dataplane.c

```diff
@@ -0,0 +1,413 @@
+	inet_ntop(AF_INET6, &ipv6Header->src_addr, ip_str, INET6_ADDRSTRLEN);
+	struct ip4to6 *new_src_addr =
+		find_ip6to4(hash, (uint8_t *)&ipv6Header->src_addr);
+	if (NULL == new_src_addr) {
+		RTE_LOG(INFO, NAT64, "not found mapping for %s. Drop\n", ip_str
+		);
+		// Если не найдено соответствующего IPv4-адреса, дропаем пакет
+		return -1;
+	}
+
+	inet_ntop(
+		AF_INET,
+		new_src_addr->ip4,
+		ip_str + INET6_ADDRSTRLEN + 1,
+		INET_ADDRSTRLEN
+	);
+	RTE_LOG(INFO,
+		NAT64,
+		"Found mapping %s -> %s\n",
+		ip_str,
+		ip_str + INET6_ADDRSTRLEN + 1);
+
+	// calculate delta for new packet length
+	uint16_t delta = packet->transport_header.offset -
+			 packet->network_header.offset -
+			 sizeof(struct rte_ipv4_hdr);
+	RTE_LOG(INFO,
+		NAT64,
+		"transport header offset: %d, network: %d, si: %lu, delta: "
+		"%d\n",
+		packet->transport_header.offset,
+		packet->network_header.offset,
+		sizeof(struct rte_ipv4_hdr),
+		delta);
+
+	struct rte_ipv4_hdr *new_ipv4_header = rte_pktmbuf_mtod_offset(
+		mbuf,
+		struct rte_ipv4_hdr *,
+		packet->network_header.offset + delta
+	);
+
+	uint16_t payload_length = rte_be_to_cpu_16(ipv6Header->payload_len);
+
+	new_ipv4_header->version_ihl = RTE_IPV4_VHL_DEF;
+	new_ipv4_header->type_of_service =
+		(rte_be_to_cpu_32(ipv6Header->vtc_flow) >> 20) & 0xFF;
+	new_ipv4_header->total_length =
+		rte_cpu_to_be_16(payload_length + sizeof(struct rte_ipv4_hdr));
+
+	new_ipv4_header->packet_id = 0;	      // TODO: generate id
+	new_ipv4_header->fragment_offset = 0; // TODO: handle fragmentation
+	new_ipv4_header->time_to_live =
+		ipv6Header->hop_limits; // TODO: decrement ttl?
+	new_ipv4_header->next_proto_id = ipv6Header->proto;
+	new_ipv4_header->hdr_checksum = 0;
+
+	memcpy(&new_ipv4_header->src_addr, new_src_addr->ip4, sizeof(uint32_t));
+	// memcpy(&new_ipv4_header->dst_addr, new_dst_addr->ip4, sizeof(struct
+	// in_addr));
+
+	// handle ICMP, TCP, UDP
+	if (ipv6Header->proto == IPPROTO_ICMPV6) {
+		new_ipv4_header->next_proto_id = IPPROTO_ICMP;
+
+		struct icmp6_hdr *icmpHeader = rte_pktmbuf_mtod_offset(
+			mbuf,
+			struct icmp6_hdr *,
+			packet->transport_header.offset
+		);
+		int result = icmp_v6_to_v4(icmpHeader);
+		if (result) {
+			RTE_LOG(ERR, NAT64, "icmp_v6_to_v4 failed\n");
+			return result;
+		}
+	}
+	// copy l2 header
+	rte_memcpy(
+		rte_pktmbuf_mtod_offset(mbuf, char *, delta),
+		rte_pktmbuf_mtod(mbuf, char *),
+		packet->network_header.offset
+	);
+
+	// reduce packet
+	if (rte_pktmbuf_adj(mbuf, delta) == NULL) {
+		RTE_LOG(ERR, NAT64, "adjust mbuf failed. Delta: %d\n", delta);
+		return -1;
+	}
+
+	// adjust new transport header offset
+	packet->transport_header.offset =
+		packet->network_header.offset + sizeof(struct rte_ipv4_hdr);
+
+	// set ipv4 header type
+	uint16_t *next_header_type = rte_pktmbuf_mtod_offset(
+		mbuf, uint16_t *, packet->network_header.offset - 2
+	);
+	*next_header_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV4);
+
+	// Обновляем длину пакета в структуре mbuf
```

Comment:
20 is a magic constant, and why adjust these fields manually? `rte_pktmbuf_adj` and `rte_pktmbuf_prepend` handle this for us."